File: src/action_example/action_client_async/src/action_client_async.cc
/******************************************************************************
 * Copyright (c) 2022-2026 SEGAR. All Rights Reserved.
 * SPDX-License-Identifier: LicenseRef-Segar-Proprietary
 *
 * PROPRIETARY AND CONFIDENTIAL. See ./LICENSE
 * for license terms and restrictions.
 *****************************************************************************/

#include <chrono>

#include "example/action/LookUpTransform.hpp"

#include "segar/segar.h"

int main(int argc, char* argv[]) {
  RETURN_VAL_IF(!rti::segar::Init(argv[0]), EXIT_FAILURE);

  auto node = rti::segar::CreateNode("action_client_async");
  RETURN_VAL_IF(!node, EXIT_FAILURE);

  using ActionClient =
      rti::segar::action::ActionClient<example::action::LookUpTransform>;
  using GoalID = rti::segar::action::GoalID;
  using GoalStatusCode = rti::segar::action::GoalStatusCode;
  using LookUpTransform = example::action::LookUpTransform;

  ActionClient::GoalCallbacks callbacks;

  callbacks.on_feedback = [](ActionClient& /*client*/, const GoalID& goal_id,
                             const LookUpTransform::Feedback& feedback) {
    AINFO << "[Async] Feedback received: goal_id="
          << rti::segar::action::internal::GoalIDToString(goal_id)
          << ", current=" << feedback.current();
  };

  callbacks.on_result = [](ActionClient& /*client*/, const GoalID& goal_id,
                           const LookUpTransform::Result& result,
                           GoalStatusCode status) {
    AINFO << "[Async] Result received: goal_id="
          << rti::segar::action::internal::GoalIDToString(goal_id)
          << ", status=" << static_cast<int>(status)
          << ", transform=" << result.transform()
          << ", error=" << result.error();
  };

  // Register cancel callback.
  callbacks.on_cancel = [](ActionClient& /*client*/, const GoalID& goal_id,
                           rti::segar::action::CancelResponseCode code) {
    AINFO << "[Async] Cancel response: goal_id="
          << rti::segar::action::internal::GoalIDToString(goal_id)
          << ", code=" << static_cast<int>(code);
  };

  auto client =
      node->CreateActionClient<LookUpTransform>("lookup_transform", callbacks);
  RETURN_VAL_IF(!client, EXIT_FAILURE);

  uint32_t index = 0;
  auto timer_callback = [&client, &index]() {
    ++index;
    LookUpTransform::Goal goal;
    goal.target_frame("map" + std::to_string(index));
    GoalID goal_id;
    if (!client->AsyncSendGoal(goal, &goal_id)) {
      AERROR << "[Async] SendGoal failed at iteration " << index;
      return;
    }
    AINFO << "[Async] Goal sent: iteration=" << index << ", goal_id="
          << rti::segar::action::internal::GoalIDToString(goal_id);

    // Cancel the current goal at a fixed interval.
    if (index > 0 && index % 5 == 0) {
      // Wait briefly so feedback has a chance to be received.
      rti::segar::SleepFor(std::chrono::milliseconds(30));
      AINFO << "[Async] Cancelling goal: goal_id="
            << rti::segar::action::internal::GoalIDToString(goal_id);
      client->AsyncCancelGoal(goal_id);
      return;
    }
  };

  // 1hz
  auto timer = std::make_shared<rti::segar::Timer>(1000, timer_callback, false);
  timer->Start();
  rti::segar::WaitForShutdown();

  return EXIT_SUCCESS;
}

File: src_python/action_example/action_client_async/src/action_client_async.py
#!/usr/bin/env python3

import time

from example.action import LookUpTransform
from segar.python import segar


def main():
    if not segar.Init("action_client_async"):
        return 1
    node = segar.CreateNode("action_client_async")
    if node is None:
        return 1

    callbacks = segar.ActionClientCallbacks()

    def on_feedback(_client, goal_id, feedback):
        print(f"[action_client_async] feedback goal={goal_id.hex()} current={feedback.current}")

    def on_result(_client, goal_id, result, status):
        print(
            f"[action_client_async] result goal={goal_id.hex()} status={status} "
            f"transform={result.transform} error={result.error}"
        )

    def on_cancel(_client, goal_id, code):
        print(f"[action_client_async] cancel response goal={goal_id.hex()} code={int(code)}")

    callbacks.on_feedback = on_feedback
    callbacks.on_result = on_result
    callbacks.on_cancel = on_cancel

    client = node.CreateActionClient("lookup_transform", LookUpTransform, callbacks=callbacks)
    if client is None:
        print("[action_client_async] create action client failed")
        return 1

    if not client.WaitForServer(3.0):
        print("[action_client_async] wait for server timeout")
        return 1

    index = 0
    while segar.OK():
        index += 1
        goal = LookUpTransform.Goal(target_frame=f"map{index}")
        goal_id = []
        if not client.AsyncSendGoal(goal, goal_id):
            print(f"[action_client_async] AsyncSendGoal failed index={index}")
            time.sleep(1.0)
            continue

        gid = goal_id[0]
        print(f"[action_client_async] goal sent index={index} goal={gid.hex()}")
        if index % 5 == 0:
            time.sleep(0.03)
            if not client.AsyncCancelGoal(gid):
                print(f"[action_client_async] AsyncCancelGoal failed goal={gid.hex()}")
            else:
                print(f"[action_client_async] cancel sent goal={gid.hex()}")

        time.sleep(1.0)

    segar.WaitForShutdown()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
